src/TCPClient.hs

```diff
@@ -46,26 +49,21 @@ import Retry
 
 -- We use parseURI rather than just splitting on : because IPv6 literals
 parsePeerName :: Int -> String -> (Int, String)
-parsePeerName defaultPort name =
-  case parseURI ("bejnarkli://" ++ name) of
-    Just uri
-      | uriScheme uri == "bejnarkli:" &&
-          uriPath uri == "" && uriQuery uri == "" && uriFragment uri == "" ->
-        case uriAuthority uri of
-          Just auth
-            | uriUserInfo auth == "" ->
-              case stripPrefix ":" (uriPort auth) of
-                Just portString ->
-                  case readMaybe portString :: Maybe Int of
-                    Just port -> (port, uriRegName auth)
-                    _ -> (defaultPort, name)
-                _ -> (defaultPort, name)
-          _ -> (defaultPort, name)
-    _ -> (defaultPort, name)
+parsePeerName defaultPort name = fromMaybe (defaultPort, name) do
+  uri <- find validURI (parseURI $ "bejnarkli://" ++ name)
+  auth <- find (null . uriUserInfo) $ uriAuthority uri
+  port <- readMaybe =<< stripPrefix ":" ( uriPort auth)
+  pure (port, uriRegName auth)
+
+  where
+    validURI uri = uriScheme uri == "bejnarkli:"
+      && null (uriPath uri)
+      && null (uriQuery uri)
+      && null (uriFragment uri)
```

Comment:
I definitely get a code smell when I see a bunch of nested case statement like this, especially when in this case it's a pyramid where all the `Nothing` cases have the same value.

I think this refactor makes the logic much easier to follow. Also note the use of `find` and `null :: Foldable t => t a -> Bool`. Gotta love those general purpose `Foldable` functions.

Reply:
This is probably my favorite refactor of my review.